### PR TITLE
[MAINT] Fixing Unit Tests for Python 3.11 and 3.12 support

### PR DESCRIPTION
**Pull request by kmilo9999**
For now we decided to test the 2 latest versions and oldest supported version of python on mac, Linux, and windows for all PRs. A full-matrix test of all supported python versions performed at a longer frequency is still being considered. 

This PR is aimed to fix incompatibilities found in the tests against python 3.11 and 3.12

### CODE REVIEW

**Comment by kmilo9999** [comment]
@gtdang @ntolley @jasmainak @dylansdaniels 
I found two problems that are making the tests fail in python 3.11+:

1. `hnn_core/tests/test_drives.py` 

The code is looking to raise a `ValueError `exception and match the error message against a specific string:
```
with pytest.raises(ValueError,
                           match=r'Burst duration (?s).* cannot be greater than'):
```
It seems that the regex rules changed in python 3.11+. It is enforced that global flags like `(?s)` need to be placed at the begging of the pattern. Moving the flag to the indicated position fixes the test case.

2. `hnn_core/tests/test_extracellular.py `

Similar to the fist point, The code is looking to raise an `AttributeError` exception and match the error message against a specific string:
```
with pytest.raises(AttributeError, match="can't set attribute"):
        rec_arr.times = [1, 2, 3]
```
In this case, pythom 3.11+ changed the message of the `AttributeError` exception to  `"property X of object Y has no setter"`, thus the test fails matching the two strings. I couldnt find this change in the python changelog, but I found the [PR ](https://github.com/python/cpython/pull/31311) that applied it.

Looking forwards for your comments.

**Comment by jasmainak** [comment]
I'm not a regular expression expert but I trust your judgement. Can you run the CIs so I can see that they get fixed?

**Comment by kmilo9999** [comment]
> I'm not a regular expression expert but I trust your judgement. Can you run the CIs so I can see that they get fixed?

Do you mean the CI in the master branch? or in this PR?

**Comment by kmilo9999** [comment]
@ntolley @gtdang It seems the unit tests are working now for python 3.11 and 3.12

File: hnn_core/tests/test_extracellular.py [hunk 1]
```diff
@@ -41,9 +41,11 @@ def test_extracellular_api():
     # all remaining input arguments checked by ExtracellularArray
 
     rec_arr = ExtracellularArray(electrode_pos)
-    with pytest.raises(AttributeError, match="can't set attribute"):
+    with pytest.raises(AttributeError,
+                       match="has no setter|can't set attribute"):
```
Comment (gtdang) [inline hnn_core/tests/test_extracellular.py:45]: Nice! Thanks for figuring out this OR condition. Could you add a comment on why we use an OR here. Maybe something about how >=3.11 changed the error msg to "has no setter".

**Review by gtdang — APPROVED** [review]
Have one minor suggestion but otherwise approve.

File: .github/workflows/unix_unit_tests.yml [hunk 2]
```diff
@@ -13,7 +13,7 @@
       strategy:
         matrix:
           os: [ubuntu-latest,macos-latest]
-          python-version: [3.8, 3.9, '3.10']
+          python-version: [3.8, 3.9, '3.10', 3.11, 3.12]
```
Comment (ntolley) [inline .github/workflows/unix_unit_tests.yml:16]:
```suggestion
          python-version: [3.8, 3.11, 3.12]
```

Really great work getting this working @kmilo9999!

As per our discussion, I think these will be the versions we stick with for now. Definitely will be keeping in mind the idea of having a separate branch for testing the full matrix in the future

Comment (gtdang) [reply .github/workflows/unix_unit_tests.yml:16]: Ah nice catch. I was discussing with a colleague that one possibility is to have a separate full-matrix workflow that tests the main branch as a cron job.  You can set up a workflow to run on a monthly 2-week basis or something.

Comment (kmilo9999) [reply .github/workflows/unix_unit_tests.yml:16]: Should I remove `3.9 `and `3.10` from the windows workflow too?

Comment (gtdang) [reply .github/workflows/unix_unit_tests.yml:16]: Yes

File: hnn_core/tests/test_extracellular.py (hunk 1 above)
Comment (kmilo9999) [reply hnn_core/tests/test_extracellular.py:48]: Done

File: .github/workflows/unix_unit_tests.yml (hunk 2 above)
Comment (kmilo9999) [reply .github/workflows/unix_unit_tests.yml]: Done

**Review by gtdang — APPROVED** [review]
Looks great!

File: .github/workflows/unix_unit_tests.yml [hunk 3]
```diff
@@ -13,7 +13,7 @@
       strategy:
         matrix:
           os: [ubuntu-latest,macos-latest]
-          python-version: [3.8, 3.9, '3.10', 3.11, 3.12]
+          python-version: [3.8, 3.11, 3.12]
```
Comment (jasmainak) [inline .github/workflows/unix_unit_tests.yml:16]: could you document the reason for this change on github so future developers know why it was done?

Comment (gtdang) [reply .github/workflows/unix_unit_tests.yml:16]: Should we document it at the top of this PR?

Comment (jasmainak) [reply .github/workflows/unix_unit_tests.yml:16]: doesn't matter as long as it's somewhere on github! If a decision was made during a private meeting, it's generally good practice to document the reasoning on github

**Comment by gtdang** [comment]
@ntolley 
The branch protections need to be updated for the versions we are testing. It's still looking for 3.9, which is no longer in the matrix. 3.11 and 3.12 should be added.

**Comment by ntolley** [comment]
All set! Thanks @kmilo9999 we're all breathing much easier with the return of the green checkmarks :smile:

**Comment by dylansdaniels** [comment]
✅ 🥳 ✅ 🥳  ✅

**Pull request merged**